applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_FIC_cut_fem_element.cpp

```diff
@@ -0,0 +1,1413 @@
+    {
+        rProjectedStress[0] = rVoigtStress[0] * rUnitNormal[0] + rVoigtStress[2] * rUnitNormal[1];
+        rProjectedStress[1] = rVoigtStress[2] * rUnitNormal[0] + rVoigtStress[1] * rUnitNormal[1];
+    }
+
+    template <>
+    void TwoStepUpdatedLagrangianVPImplicitFluidFicCutFemElement<3>::VoigtStressNormalProjection(
+        const Vector &rVoigtStress,
+        const array_1d<double, 3> &rUnitNormal,
+        array_1d<double, 3> &rProjectedStress)
+    {
+        rProjectedStress[0] = rVoigtStress[0] * rUnitNormal[0] + rVoigtStress[3] * rUnitNormal[1] + rVoigtStress[5] * rUnitNormal[2];
+        rProjectedStress[1] = rVoigtStress[3] * rUnitNormal[0] + rVoigtStress[1] * rUnitNormal[1] + rVoigtStress[4] * rUnitNormal[2];
+        rProjectedStress[2] = rVoigtStress[5] * rUnitNormal[0] + rVoigtStress[4] * rUnitNormal[1] + rVoigtStress[2] * rUnitNormal[2];
+    }
+
+    template <>
+    void TwoStepUpdatedLagrangianVPImplicitFluidFicCutFemElement<2>::CalculateBMatrix(
+        const Matrix &rDNDX,
+        BoundedMatrix<double, StrainSize, 2 * NumNodes> &rB)
+    {
+        IndexType index;
+        rB = ZeroMatrix(StrainSize, 2 * NumNodes);
+        for (unsigned int i = 0; i < NumNodes; i++)
+        {
+            index = 2 * i;
+            rB(0, index + 0) = rDNDX(i, 0);
+            rB(1, index + 1) = rDNDX(i, 1);
+            rB(2, index + 0) = rDNDX(i, 1);
+            rB(2, index + 1) = rDNDX(i, 0);
+        }
+    }
+
+    template <>
+    void TwoStepUpdatedLagrangianVPImplicitFluidFicCutFemElement<3>::CalculateBMatrix(
+        const Matrix &rDNDX,
+        BoundedMatrix<double, StrainSize, 3 * NumNodes> &rB)
+    {
+        IndexType index;
+        rB = ZeroMatrix(StrainSize, 3 * NumNodes);
+        for (unsigned int i = 0; i < NumNodes; i++)
+        {
+            index = 3 * i;
+            rB(0, index + 0) = rDNDX(i, 0);
+            rB(1, index + 1) = rDNDX(i, 1);
+            rB(2, index + 2) = rDNDX(i, 2);
+            rB(3, index + 0) = rDNDX(i, 1);
+            rB(3, index + 1) = rDNDX(i, 0);
+            rB(4, index + 1) = rDNDX(i, 2);
+            rB(4, index + 2) = rDNDX(i, 1);
+            rB(5, index + 0) = rDNDX(i, 2);
+            rB(5, index + 2) = rDNDX(i, 0);
+        }
+    }
+
+    template <>
+    void TwoStepUpdatedLagrangianVPImplicitFluidFicCutFemElement<2>::VoigtTransformForProduct(
+        const array_1d<double, 3> &rVector,
+        BoundedMatrix<double, 2, StrainSize> &rVoigtMatrix)
+    {
+
+        rVoigtMatrix.clear();
+
+        rVoigtMatrix(0, 0) = rVector(0);
+        rVoigtMatrix(0, 2) = rVector(1);
+        rVoigtMatrix(1, 1) = rVector(1);
+        rVoigtMatrix(1, 2) = rVector(0);
+    }
+
+    template <>
+    void TwoStepUpdatedLagrangianVPImplicitFluidFicCutFemElement<3>::VoigtTransformForProduct(
+        const array_1d<double, 3> &rVector,
+        BoundedMatrix<double, 3, StrainSize> &rVoigtMatrix)
+    {
+
+        rVoigtMatrix.clear();
+
+        rVoigtMatrix(0, 0) = rVector(0);
```

Comment:
I imagine that you are not transforming strains... Anyway this is implemented in MathUtils like: `MathUtils<double>::StressVectorToTensor(matrix_stress_vector)` or `MathUtils<double>::StrainVectorToTensor(matrix_stress_vector)`